libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * AVL Tree implementation.
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan 
+{
+template <typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable 
+{
+protected:
+    struct Node {
+        T* data = UAVCAN_NULLPTR;
+        int16_t h = 1; // initially added as leaf
+        Node* left = UAVCAN_NULLPTR;
+        Node* right = UAVCAN_NULLPTR;
+        Node* equal_keys = UAVCAN_NULLPTR;
+    };
+
+    Node* root_;
+
+private:
+    size_t len_ ;
+
+    static int16_t heightOf(const Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return n->h;
+    }
+
+    void postOrderTraverseRecursively(Node* n, std::function<void(T*&)> forEach) {
+        if (n == UAVCAN_NULLPTR) {
+            return;
+        }
+
+        postOrderTraverseRecursively(n->left, forEach);
+        postOrderTraverseRecursively(n->right, forEach);
+        forEach(n->data);
+    }
+
+    Node* makeNode(T* payload) {
+        void* praw = this->allocator_.allocate(sizeof(Node));
+
+        if (praw == UAVCAN_NULLPTR) {
+            UAVCAN_TRACE("AvlTree", " OOM -- Can't allocate Node");
+            return UAVCAN_NULLPTR; // Push rejected
+        }
+
+        Node* node = new (praw) Node();
+        UAVCAN_ASSERT(node);
+
+        node->data = payload;
+        return node;
+    }
+
+    void deleteNode(Node*& n) {
+        if (n != UAVCAN_NULLPTR) {
+            n->~Node();
+            allocator_.deallocate(n);
+            n = UAVCAN_NULLPTR;
+        }
+    }
+
+    static int16_t balanceOf(Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return static_cast<int16_t>(heightOf(n->left) - heightOf(n->right));
+    }
+
+    static int16_t maxOf(int16_t a, int16_t b) {
+        return static_cast<int16_t>(a > b ? a : b);
+    }
+
+    static Node* rotateRight(Node* y) {
+        Node* x = y->left;
+        Node* T2 = x->right;
+
+        x->right = y;
+        y->left = T2;
+
+        y->h = static_cast<int16_t>(maxOf(heightOf(y->left), heightOf(y->right)) + 1);
```

Comment:
Almost all the uses of `static_cast<int16_t>` in this implementation are at best redundant and at worst a code maintainability issue (i.e. Casts can suppress compiler warnings about integer size mismatch or prevent implicit promotions if integer types are changed). Please only use a cast where it is required.

Reply:
`-Werror=conversion` exists and it wouldn't compile otherwise; On every usage of a static cast that's not in `heightOf` , `maxOf`, it happens, because there is a possible overflow if you increment the returned `int16_t`.

`maxOf(heightOf(y->left), heightOf(y->right)) + 1`

Reply:
yep you are right. You should probably just ignore everything I say at this point.